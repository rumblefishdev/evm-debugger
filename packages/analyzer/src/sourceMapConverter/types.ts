import type { TSourceMap } from '@evm-debuger/types'

import type { Opcodes } from './opcodes'

export type JsonContent = {
  sourceMaps: Record<string, TSourceMap[]>
  sourceCodes: Record<string, string>
}

export type Source = {
  bytecode: string
  sourceMap: string
  opcodes: string
  sourceCodeRaw: string
  sourceCodeParsed: string
}

export type PayloadContent = {
  address: string
  combinedRawSourceCode: string
  sources: Record<string, Source>
}

export type OpcodeElement = {
  opcode: Opcodes | string
  payload: string
}

export type SourceMapElement = {
  start: number
  length: number
  fileId: number
  jump: string
}

export type ConversionResult = {
  pc: number
  opcode: Opcodes
  fileId: number
  fileLine: number
}

export type Payload = Record<string, PayloadContent>
