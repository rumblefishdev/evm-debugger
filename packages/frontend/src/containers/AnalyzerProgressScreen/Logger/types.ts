import type { StackProps } from '@mui/material'

import type { TLogMessageRecord } from '../../../store/analyzer/analyzer.types'

export interface LoggerProps extends StackProps {
  messages: TLogMessageRecord[]
}
