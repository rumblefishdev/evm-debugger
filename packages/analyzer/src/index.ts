export { TxAnalyzer } from './txAnalyzer'
export {
  checkOpcodeIfOfCallGroupType,
  checkOpcodeIfOfCallOrStaticType,
  checkOpcodeIfOfCreateGroupType,
  checkOpcodeIfOfDelegateCallType,
  checkOpcodeIfOfFunctionBlockEndType,
  checkOpcodeIfOfLogGroupType,
  checkOpcodeIfOfReturnGroupType,
} from './helpers/structLogTypeGuards'
export { Opcodes, AlternativeOpcodes } from './opcodes/opcodes'
