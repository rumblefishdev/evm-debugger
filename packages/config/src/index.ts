import dotenv from 'dotenv'

dotenv.config()

export * from './chains'
export * from './config'
export * from './utils'
