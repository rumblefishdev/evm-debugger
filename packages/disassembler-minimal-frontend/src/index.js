import { Disassembler } from './disassembler'

const disassembler = new Disassembler()

// document.getElementById('form').addEventListener('submit', (ev) => {
//   ev.preventDefault();
//   const hexcode = document.getElementById('code').value;
//   disassembler.disassemble(hexcode).then((result) => {
//     document.getElementById('result').textContent = JSON.stringify(result, null, 2);
//   });
// })
