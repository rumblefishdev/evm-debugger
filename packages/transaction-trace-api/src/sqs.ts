import { SendMessageCommand, SQSClient } from '@aws-sdk/client-sqs'
import { forkingUrlMap } from '@evm-debuger/types'

const sqsClient = new SQSClient({ region: process.env.AWS_REGION })

export const putTxDetailsToSqs = async (
  txHash: string,
  chainId: string,
  gasLimit: string,
) => {
  const params = {
    QueueUrl: process.env[`SQS_ANALYZER_URL_${chainId}`],
    MessageBody: `Transaction ${txHash}`,
    MessageAttributes: {
      txHash: {
        StringValue: txHash,
        DataType: 'String',
      },
      hardhatForkingUrl: {
        StringValue:
          forkingUrlMap[Number(chainId) as keyof typeof forkingUrlMap],
        DataType: 'String',
      },
      gasLimit: {
        StringValue: gasLimit,
        DataType: 'String',
      },
      chainId: {
        StringValue: chainId,
        DataType: 'String',
      },
    },
    DelaySeconds: 0,
  }
  console.log(JSON.stringify(params))
  const command = new SendMessageCommand(params)
  await sqsClient.send(command)
}
