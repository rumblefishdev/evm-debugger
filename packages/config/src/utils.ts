import { ALCHEMY_API_URLS, BLOCKCHAIN_API_URLS, ChainIds } from './chains'
import { API_KEYS } from './config'

export const getAlchemyApiUrl = (chainId: ChainIds): string => {
  return `${ALCHEMY_API_URLS[chainId]}${API_KEYS.alchemy}`
}

export const getBlockchainApiUrl = (chainId: ChainIds): { blockchainApiUrl: string; blockchainApiKey: string } => {
  switch (chainId) {
    case ChainIds.ETHEREUM_MAINNET:
    case ChainIds.ETHEREUM_GOERLI:
    case ChainIds.ETHEREUM_SEPOLIA:
      return { blockchainApiUrl: `${BLOCKCHAIN_API_URLS[chainId]}/api/`, blockchainApiKey: API_KEYS.ethereum }
    case ChainIds.POLYGON_MAINNET:
    case ChainIds.POLYGON_MUMBAI:
      return { blockchainApiUrl: `${BLOCKCHAIN_API_URLS[chainId]}/api/`, blockchainApiKey: API_KEYS.polygon }
    case ChainIds.ARBITRUM_MAINNET:
    case ChainIds.ARBITRUM_GOERLI:
      return { blockchainApiUrl: `${BLOCKCHAIN_API_URLS[chainId]}/api/`, blockchainApiKey: API_KEYS.arbitrum }
    default:
      return { blockchainApiUrl: '', blockchainApiKey: '' }
  }
}
