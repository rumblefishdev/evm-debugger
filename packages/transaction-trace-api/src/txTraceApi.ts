import type { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda'
import { TransactionTraceResponseStatus } from '@evm-debuger/types'
import { AWSLambda, captureException } from '@sentry/serverless'

import { version } from '../package.json'

import {
  getTransactionDetails,
  putTxDetailsToDdb,
  putTxEventToDdb,
} from './ddb'
import { putTxDetailsToSqs } from './sqs'

AWSLambda.init({
  tracesSampleRate: 1,
  release: version,
  environment: process.env.ENVIRONMENT,
  dsn: process.env.SENTRY_DSN,
})
AWSLambda.setTag('lambda_name', 'transaction-trace-api')

export const createResponse = (status: string, output = {}) => {
  return {
    statusCode: 200,
    headers: {
      'Access-Control-Allow-Origin': '*',
    },
    body: JSON.stringify({
      status,
      ...output,
    }),
  }
}

export const analyzeTransactionHandler = async (
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
  const { txHash, chainId } = event.pathParameters!
  if (!txHash || !chainId)
    return createResponse(
      TransactionTraceResponseStatus.FAILED,
      'Invalid params',
    )
  console.log('rafadddslddd')
  try {
    let txDetails = await getTransactionDetails(txHash)

    if (txDetails === null) {
      txDetails = await putTxDetailsToDdb(txHash, chainId)
      await putTxDetailsToSqs(txHash, chainId)
    }

    if (txDetails.status === TransactionTraceResponseStatus.FAILED) {
      await putTxEventToDdb(TransactionTraceResponseStatus.PENDING, txHash)
      await putTxDetailsToSqs(txHash, chainId)
    }
    return createResponse(TransactionTraceResponseStatus.SUCCESS, txDetails)
  } catch (error) {
    if (error instanceof Error) {
      console.log(error)
      captureException(error)
      return createResponse(TransactionTraceResponseStatus.FAILED)
    }
  }
  return createResponse(TransactionTraceResponseStatus.FAILED)
}

export const analyzeTransactionEntrypoint = AWSLambda.wrapHandler(
  analyzeTransactionHandler,
)
