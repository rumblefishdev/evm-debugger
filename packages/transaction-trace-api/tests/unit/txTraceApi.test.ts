import type { APIGatewayProxyResult, Context } from 'aws-lambda'
import { mockClient } from 'aws-sdk-client-mock'
import { DynamoDBDocumentClient, QueryCommand } from '@aws-sdk/lib-dynamodb'
import { SQSClient } from '@aws-sdk/client-sqs'
import timekeeper from 'timekeeper'
import { TransactionTraceResponseStatus } from '@evm-debuger/types'

import { createLambdaEvent } from '../utils/lambdaMocks'
import {
  getMockCalledInput,
  getMockCalledInputItem,
} from '../utils/awsMocksHelper'
import { analyzeTransactionHandler } from '../../src/txTraceApi'

const ddbMock = mockClient(DynamoDBDocumentClient)
const sqsMock = mockClient(SQSClient)

describe('Unit test for api', function () {
  const TX_HASH =
    '0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9'
  const CHAIN_ID = '1'
  const txInitDetails = {
    txHash: TX_HASH,
    chainId: CHAIN_ID,
  }
  const testEvent = createLambdaEvent(txInitDetails)

  // TODO: Mock this context in better way
  // Do not use aws-lambda-mock-context because it is not maintained
  const ctx: Context = {
    succeed: () => {},
    memoryLimitInMB: '128',
    logStreamName: '2014/02/14/[HEAD]13370a84ca4ed8b77c427af260',
    logGroupName: 'transaction-trace-api',
    invokedFunctionArn:
      'arn:aws:lambda:us-east-1:123456789012:function:transaction-trace-api',
    getRemainingTimeInMillis: () => 1000,
    functionVersion: '1',
    functionName: 'transaction-trace-api',
    fail: () => {},
    done: () => {},
    callbackWaitsForEmptyEventLoop: true,
    awsRequestId: '123',
  }

  beforeAll(() => {
    timekeeper.freeze(new Date('2014-01-01'))
  })
  beforeEach(() => {
    ddbMock.reset()
    sqsMock.reset()
  })
  afterAll(() => {
    timekeeper.reset()
  })

  it('create analyze if 1st time', async () => {
    ddbMock.on(QueryCommand).resolves({ Items: [] })

    const result: APIGatewayProxyResult = await analyzeTransactionHandler(
      testEvent,
      ctx,
    )
    expect(ddbMock.calls().length).toEqual(2)
    expect(sqsMock.calls().length).toEqual(1)

    const putTxInfoInput = getMockCalledInputItem(ddbMock, 1)
    expect(putTxInfoInput.txHash).toEqual(TX_HASH)
    expect(putTxInfoInput.chainId).toEqual(CHAIN_ID)
    expect(putTxInfoInput.status).toEqual(
      TransactionTraceResponseStatus.PENDING,
    )
    expect(putTxInfoInput['type#time']).toEqual('TRANSACTION')

    expect(getMockCalledInput(sqsMock, 0)).toMatchSnapshot()

    expect(result.statusCode).toEqual(200)
    const txResponse = JSON.parse(result.body)
    expect(txResponse).toMatchSnapshot()
  })

  it('already in progress, return status', async () => {
    ddbMock.on(QueryCommand).resolves({
      Items: [
        {
          ...txInitDetails,
        },
        {
          txHash: TX_HASH,
          status: TransactionTraceResponseStatus.RUNNING,
        },
      ],
    })

    const result: APIGatewayProxyResult = await analyzeTransactionHandler(
      testEvent,
      ctx,
    )
    expect(ddbMock.calls().length).toEqual(1)
    expect(sqsMock.calls().length).toEqual(0)

    expect(result.statusCode).toEqual(200)
    const txResponse = JSON.parse(result.body)
    expect(txResponse).toMatchSnapshot()
  })

  it('rerun analyze on failed tx', async () => {
    ddbMock.on(QueryCommand).resolves({
      Items: [
        {
          ...txInitDetails,
        },
        {
          txHash: TX_HASH,
          status: TransactionTraceResponseStatus.FAILED,
        },
        {
          txHash: TX_HASH,
          status: TransactionTraceResponseStatus.RUNNING,
        },
      ],
    })

    const result: APIGatewayProxyResult = await analyzeTransactionHandler(
      testEvent,
      ctx,
    )
    expect(ddbMock.calls().length).toEqual(2)
    expect(sqsMock.calls().length).toEqual(1)

    const putTxEventInput = getMockCalledInputItem(ddbMock, 1)
    expect(putTxEventInput.status).toEqual(
      TransactionTraceResponseStatus.PENDING,
    )

    expect(result.statusCode).toEqual(200)
    const txResponse = JSON.parse(result.body)
    expect(txResponse).toMatchSnapshot()
  })

  it('analyze success', async () => {
    ddbMock.on(QueryCommand).resolves({
      Items: [
        {
          txHash: TX_HASH,
          status: TransactionTraceResponseStatus.PENDING,
          s3Location: `trace/${CHAIN_ID}/${TX_HASH}`,
          chainId: CHAIN_ID,
        },
        {
          txHash: TX_HASH,
          status: TransactionTraceResponseStatus.SUCCESS,
        },
        {
          txHash: TX_HASH,
          status: TransactionTraceResponseStatus.RUNNING,
        },
      ],
    })

    const result: APIGatewayProxyResult = await analyzeTransactionHandler(
      testEvent,
      ctx,
    )
    expect(ddbMock.calls().length).toEqual(1)
    expect(sqsMock.calls().length).toEqual(0)

    expect(result.statusCode).toEqual(200)
    const txResponse = JSON.parse(result.body)
    expect(txResponse).toMatchSnapshot()
  })
})
