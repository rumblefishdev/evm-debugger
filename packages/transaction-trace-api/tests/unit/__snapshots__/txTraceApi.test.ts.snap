// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for api already in progress, return status 1`] = `
{
  "chainId": "1",
  "status": "RUNNING",
  "txHash": "0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9",
}
`;

exports[`Unit test for api analyze success 1`] = `
{
  "chainId": "1",
  "s3Location": "trace/1/0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9",
  "status": "SUCCESS",
  "txHash": "0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9",
}
`;

exports[`Unit test for api create analyze if 1st time 1`] = `
{
  "DelaySeconds": 0,
  "MessageAttributes": {
    "chainId": {
      "DataType": "String",
      "StringValue": "1",
    },
    "hardhatForkingUrl": {
      "DataType": "String",
      "StringValue": "https://eth-mainnet.alchemyapi.io/v2/",
    },
    "txHash": {
      "DataType": "String",
      "StringValue": "0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9",
    },
  },
  "MessageBody": "Transaction 0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9",
  "QueueUrl": undefined,
}
`;

exports[`Unit test for api create analyze if 1st time 2`] = `
{
  "chainId": "1",
  "status": "PENDING",
  "timestamp": "1388534400000",
  "txHash": "0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9",
  "type#time": "TRANSACTION",
}
`;

exports[`Unit test for api rerun analyze on failed tx 1`] = `
{
  "chainId": "1",
  "status": "FAILED",
  "txHash": "0xf2a56c4a9edc31fd3a8ed3c3e256d500f548035e84e55df6e1c6b631d91c04f9",
}
`;
